instrumentation/undertow-1.4/src/main/java/io/opentelemetry/javaagent/instrumentation/hypertrace/undertow/v1_4/StreamSourceChannelInstrumentation.java

```diff
@@ -0,0 +1,89 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.opentelemetry.javaagent.instrumentation.hypertrace.undertow.v1_4;
+
+import static net.bytebuddy.matcher.ElementMatchers.isPublic;
+import static net.bytebuddy.matcher.ElementMatchers.named;
+import static net.bytebuddy.matcher.ElementMatchers.returns;
+import static net.bytebuddy.matcher.ElementMatchers.takesArgument;
+import static net.bytebuddy.matcher.ElementMatchers.takesArguments;
+
+import io.opentelemetry.javaagent.instrumentation.api.ContextStore;
+import io.opentelemetry.javaagent.instrumentation.api.InstrumentationContext;
+import io.opentelemetry.javaagent.instrumentation.hypertrace.undertow.v1_4.utils.Utils;
+import io.opentelemetry.javaagent.tooling.TypeInstrumentation;
+import io.opentelemetry.javaagent.tooling.bytebuddy.matcher.AgentElementMatchers;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import net.bytebuddy.matcher.ElementMatcher.Junction;
+import org.hypertrace.agent.core.instrumentation.HypertraceCallDepthThreadLocalMap;
+import org.hypertrace.agent.core.instrumentation.SpanAndObjectPair;
+import org.xnio.channels.StreamSourceChannel;
+
+public final class StreamSourceChannelInstrumentation implements TypeInstrumentation {
+
+  @Override
+  public ElementMatcher<TypeDescription> typeMatcher() {
+    return AgentElementMatchers.safeHasSuperType(named("org.xnio.channels.StreamSourceChannel"));
+  }
+
+  @Override
+  public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+    final Map<Junction<MethodDescription>, String> matchers = new HashMap<>();
+    matchers.put(
+        named("read")
+            .and(takesArguments(1))
+            .and(takesArgument(0, ByteBuffer.class))
+            .and(returns(int.class))
+            .and(isPublic()),
+        StreamSourceChannelInstrumentation.class.getName() + "$read_advice");
+    return Collections.unmodifiableMap(matchers);
+  }
+
+  /** TODO */
+  static final class read_advice {
+
+    @Advice.OnMethodEnter
+    public static int trackCallDepth() {
+      return HypertraceCallDepthThreadLocalMap.incrementCallDepth(StreamSourceChannel.class);
+    }
+
+    @Advice.OnMethodExit(suppress = Throwable.class)
+    public static void enter(
+        @Advice.Enter int callDepth,
+        @Advice.Return final int numBytesRead,
+        @Advice.This final StreamSourceChannel streamSourceChannel,
+        @Advice.Argument(0) final ByteBuffer byteBuffer) {
+      if (callDepth > 0) {
+        return;
+      }
+      HypertraceCallDepthThreadLocalMap.reset(StreamSourceChannel.class);
```

Comment:
if an exception is thrown in the method the thread map will not be reset.

Reply:
Thanks! Good catch 🎉